node/src/components/consensus/utils.rs

```diff
@@ -273,4 +300,191 @@ mod tests {
+            &insufficient,
+            custom_quorum_formula,
+        );
+        assert!(matches!(
+            result,
+            Err(FinalitySignatureError::InsufficientWeightForFinality {
+                trusted_validator_weights: _,
+                block_signatures: _,
+                signature_weight,
+                total_validator_weight,
+                finality_threshold_fraction: _
+            }) if *total_validator_weight == TOTAL_VALIDATORS_WEIGHT.into() && *signature_weight == INSUFFICIENT_FINALITY_SIGNATURES.into()
+        ));
+
+        let result = check_sufficient_finality_signatures_with_quorum_formula(
+            &validator_weights,
+            finality_threshold_fraction,
+            &just_enough_weight,
+            custom_quorum_formula,
+        );
+        assert!(result.is_ok());
+
+        let result = check_sufficient_finality_signatures_with_quorum_formula(
+            &validator_weights,
+            finality_threshold_fraction,
+            &too_many,
+            custom_quorum_formula,
+        );
+        assert!(matches!(
+            result,
+            Err(FinalitySignatureError::TooManySignatures {
+                trusted_validator_weights: _,
+                block_signatures: _,
+                signature_weight,
+                weight_minus_minimum: _,
+                total_validator_weight,
+                finality_threshold_fraction: _
+            }) if *total_validator_weight == TOTAL_VALIDATORS_WEIGHT.into() && *signature_weight == TOO_MANY_FINALITY_SIGNATURES.into()
+        ));
+    }
+
+    #[test]
+    fn detects_bogus_validator() {
+        const TOTAL_VALIDATORS: usize = 20;
+        const JUST_ENOUGH_FINALITY_SIGNATURES: usize = 14;
+
+        let mut rng = TestRng::new();
+
+        let (validators, validator_weights) = generate_validators(TOTAL_VALIDATORS);
+        let finality_threshold_fraction = Ratio::new_raw(1, 3);
+
+        // Generate correct signatures.
+        let mut signatures =
+            create_signatures(&mut rng, &validators, JUST_ENOUGH_FINALITY_SIGNATURES);
+        let result = check_sufficient_finality_signatures(
+            &validator_weights,
+            finality_threshold_fraction,
+            &signatures,
+        );
+        assert!(result.is_ok());
+
+        // Smuggle a bogus proof in.
+        let (_, pub_key) = generate_ed25519_keypair();
+        signatures.insert_proof(pub_key.clone(), *signatures.proofs.iter().next().unwrap().1);
+        let result = check_sufficient_finality_signatures(
+            &validator_weights,
+            finality_threshold_fraction,
+            &signatures,
+        );
+        assert!(matches!(
+            result,
+            Err(FinalitySignatureError::BogusValidator {
+                trusted_validator_weights: _,
+                block_signatures: _,
+                bogus_validator_public_key
+            })  if *bogus_validator_public_key == pub_key
+        ));
```

Comment:
I thought we already have a test like that :)

Reply:
Yes, there is similar one.

This one is exercising the finality signatures sufficiency function, which can return `BogusValidator` as a side effect of the weight calculation.

The other test is exercising the `validate_finality_signatures` which is invoked on the signatures fetched from peer.

This is not perfect, but simple approach.

We could use some deduplication here, but ideally, I'd like the `validate_finality_signatures` to return a new type (`ValidatedFinalitySignatures`) and then we can trust there is no bogus validator at the calculation stage, hence making the code a bit more bullet-proof. But that might be a subject to a bigger refactor.

cc: @marc-casperlabs  since we discussed similar approach already